File: lib/features/product/domain/usecases/product_usecases.dart
import 'package:dartz/dartz.dart';

import '../../../../core/errors/failures.dart';
import '../../../../core/usecase/usecase.dart';
import '../entities/product_entity.dart';
import '../repositories/product_repository.dart';

class GetAllProducts extends UseCase<List<ProductEntity>, NoParams> {
  const GetAllProducts(this._repo);
  final ProductRepository _repo;

  @override
  Future<Either<Failure, List<ProductEntity>>> call(NoParams params) =>
      _repo.getAll();
}

class GetProductById extends UseCase<ProductEntity, String> {
  const GetProductById(this._repo);
  final ProductRepository _repo;

  @override
  Future<Either<Failure, ProductEntity>> call(String id) => _repo.getById(id);
}

class AddProduct extends UseCase<ProductEntity, ProductEntity> {
  const AddProduct(this._repo);
  final ProductRepository _repo;

  @override
  Future<Either<Failure, ProductEntity>> call(ProductEntity params) =>
      _repo.add(params);
}

class UpdateProduct extends UseCase<ProductEntity, ProductEntity> {
  const UpdateProduct(this._repo);
  final ProductRepository _repo;

  @override
  Future<Either<Failure, ProductEntity>> call(ProductEntity params) =>
      _repo.update(params);
}

class DeleteProduct extends UseCase<Unit, String> {
  const DeleteProduct(this._repo);
  final ProductRepository _repo;

  @override
  Future<Either<Failure, Unit>> call(String id) => _repo.delete(id);
}

class SearchProducts extends UseCase<List<ProductEntity>, String> {
  const SearchProducts(this._repo);
  final ProductRepository _repo;

  @override
  Future<Either<Failure, List<ProductEntity>>> call(String query) =>
      _repo.search(query);
}

File: packages/example_react_app/src/features/product/domain/usecases/product.usecases.ts
/// Use cases for Product.
import type { Either } from '../../../../core/usecase/usecase';
import type { Failure } from '../../../../core/errors/failures';
import type { ProductEntity } from '../entities/product.entity';
import type { ProductRepository } from '../repositories/product.repository';
import type { PaginationParams, PaginatedResponse } from '../../../../core/pagination/pagination';

export class GetAllProductsUseCase {
  private repository: ProductRepository;

  constructor(repository: ProductRepository) {
    this.repository = repository;
  }

  async execute(): Promise<Either<Failure, ProductEntity[]>> {
    return this.repository.getAll();
  }
}

export class GetProductByIdUseCase {
  private repository: ProductRepository;

  constructor(repository: ProductRepository) {
    this.repository = repository;
  }

  async execute(id: string): Promise<Either<Failure, ProductEntity>> {
    return this.repository.getById(id);
  }
}

export class AddProductUseCase {
  private repository: ProductRepository;

  constructor(repository: ProductRepository) {
    this.repository = repository;
  }

  async execute(entity: ProductEntity): Promise<Either<Failure, ProductEntity>> {
    return this.repository.add(entity);
  }
}

export class UpdateProductUseCase {
  private repository: ProductRepository;

  constructor(repository: ProductRepository) {
    this.repository = repository;
  }

  async execute(entity: ProductEntity): Promise<Either<Failure, ProductEntity>> {
    return this.repository.update(entity);
  }
}

export class DeleteProductUseCase {
  private repository: ProductRepository;

  constructor(repository: ProductRepository) {
    this.repository = repository;
  }

  async execute(id: string): Promise<Either<Failure, void>> {
    return this.repository.delete(id);
  }
}

export class SearchProductsUseCase {
  private repository: ProductRepository;

  constructor(repository: ProductRepository) {
    this.repository = repository;
  }

  async execute(query: string): Promise<Either<Failure, ProductEntity[]>> {
    return this.repository.search(query);
  }
}

export class GetAllProductsPaginatedUseCase {
  private repository: ProductRepository;

  constructor(repository: ProductRepository) {
    this.repository = repository;
  }

  async execute(params: PaginationParams): Promise<Either<Failure, PaginatedResponse<ProductEntity>>> {
    return this.repository.getAllPaginated(params);
  }
}
